### ISSUE TITLE
Does ouch support permissions like `--xattrs`?

### ISSUE THREAD

**Issue by RyanGreenup** [ouch-org/ouch#626]
### Discussed in https://github.com/ouch-org/ouch/discussions/625

<div type='discussions-op-text'>

<sup>Originally posted by **RyanGreenup** February 10, 2024</sup>
It's common to use `tar` to backup up a home directory like so:

```bash
mount LABEL=ROOT -o subvol=@ /mnt

# use bsdar if gnu tar compiled without xattrs or acls
cd /mnt
x=/root-tarball.tar.gz
bsdtar --acls --xattrs --preserve-permissions -cvaf ${x} .
```

Users can later extract this with:

```bash
bsdtar --acls --xattrs  --numeric-owner -xpzf ${x} /mnt
```
Is there any support for those properties in _Ouch_? It would be convenient, atleast for me, if there was a command line flag like `-P` / `--preserve-all-permissions`, then one could more simply:

```bash
# Compress
cd /mnt
ouch compress -P . ${x}

# Decompress
cd /
ouch decompress -P ${x} .
```
</div>

I had a look at the:

* [tar crate](https://github.com/alexcrichton/tar-rs?tab=readme-ov-file)
  * See also [Tar - Extended File Attributes Docs](https://www.gnu.org/software/tar/manual/html_node/Extended-File-Attributes.html).

and the crate documentation suggests the following functions:

* `--xattrs`
  * [set_unpack_xattrs](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_unpack_xattrs)
* `--preserve-permissions`
  * [set_preserve_permissions](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_preserve_permissions)
* `--acls`
  * I couldn't find anything for this, I'll have a look through the docs, it may have a different name.
* `--numeric-owner`
  * This could be important for Docker containers
  * [set_preserve_ownerships](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_preserve_ownerships)

We'd probably need to add this somewhere around [src/archive/tar.rs](https://github.com/ouch-org/ouch/blob/0.5.1/src/archive/tar.rs#L93)

If you could offer some guidance, I can try and put together a PR if you'd like?

**Comment by RyanGreenup** [comment]
These are all the functions I could find under `Archive::` with the `set_` prefix:

* [tar::Archive::set_mask](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_mask)
* [tar::Archive::set_overwrite](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_overwrite)
* [tar::Archive::set_ignore_zeros](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_ignore_zeros)
* [tar::Archive::set_unpack_xattrs](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_unpack_xattrs)
* [tar::Archive::set_preserve_mtime](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_preserve_mtime)
* [tar::Archive::set_preserve_ownerships](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_preserve_ownerships)
* [tar::Archive::set_preserve_permissions](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_preserve_permissions)


There's also a [set_mode](https://docs.rs/tar/latest/tar/struct.Header.html#method.set_mode) which might be it.

**Comment by RyanGreenup** [comment]
It seems like the work flow of tar is:

1.  Compress
    1.  Make a
        [builder](https://docs.rs/tar/latest/tar/struct.Builder.html#method.mode)
        objectwith a
        [Header::new<sub>gnu</sub>()](https://docs.rs/tar/latest/tar/struct.Builder.html#method.append),
        - this object uses [builder.append(&mut
          header)](https://docs.rs/tar/latest/tar/struct.Builder.html#method.append)
          to add files
        - that header will by default have
          [HeaderMode::Complete](https://docs.rs/tar/latest/tar/enum.HeaderMode.html#variant.Complete)
          to include everything like mod/access times and ownership
          - (although it could be modified by it's
            [set<sub>metadata</sub>](https://docs.rs/tar/latest/tar/struct.Header.html#method.set_metadata)
            method)
            - `TODO` does this include acls by default?
    2.  Use `builder.append` to add files
2.  Decompress
    1.  Create an archive and set
        - [set<sub>unpackxattrs</sub>](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_unpack_xattrs)
        - [set<sub>preservepermissions</sub>](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_preserve_permissions)
        - [set<sub>preserveownerships</sub>](https://docs.rs/tar/latest/tar/struct.Archive.html#method.set_preserve_ownerships)
        - set<sub>acls</sub>
          - `TODO` is there a way to set acls?
    2.  Unpack the archive

For Compressing [the example in the
documentation](https://docs.rs/tar/latest/tar/struct.Builder.html#examples)
is something like:

``` rust
use tar::{Builder, Header};

let mut header = Header::new_gnu();

let mut data: &[u8] = &[1, 2, 3, 4];

let mut builder = Builder::new(Vec::new());
builder.append(&header, data).unwrap();
let data = builder.into_inner().unwrap();
```

We just need to ensure acls are pulled in.

This could then be decompressed with
ouch/src/arcive/tar.rs::unpack<sub>archive</sub>() like so:

``` rust
pub fn unpack_archive(reader: Box<dyn Read>, output_folder: &Path, quiet: bool) -> crate::Result<usize> {
    assert!(output_folder.read_dir().expect("dir exists").count() == 0);
    let mut archive = tar::Archive::new(reader);

    // --xattrs
    archive.set_unpack_xattrs(true);
    // --preserver-permissions
    archive.set_preserve_permissions();
    // --numeric-ownership
    archive.set_preserve_ownerships();

    // --acls
    // TODO

    let mut files_unpacked = 0;
    for file in archive.entries()? {
        let mut file = file?;

        file.unpack_in(output_folder)?;
    }

    Ok(files_unpacked)
}
```

Does that seem right?

list.rs:67

Minor PR

The files<sub>unpacked</sub> variable in in
[unpack<sub>archive</sub>](https://github.com/ouch-org/ouch/blob/0.5.1/src/archive/tar.rs#L46)
is only incremented if not `quiet`

this is used in decompress.rs:115 in the decompress<sub>file</sub>
function, this function is used in mod.rs:178 which takes quiet from the
args. ahhh but then it's only used in decompress<sub>files</sub> to
report the number of decompressed files. so there's no need to count if
it's quiet

The docstring should be modified.

**Comment by RyanGreenup** [comment]
I created an issue with tar-rs asking about acls:

* [tar-rs #353](https://github.com/alexcrichton/tar-rs/issues/353)

Further resources:
* [Wikipedia - Table of archive Formats supporting ACLs](https://en.wikipedia.org/wiki/List_of_archive_formats#Features)
* [RHEL - Archiving File Systems With ACLs](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/4/html/system_administration_guide/access_control_lists-archiving_file_systems_with_acls)

**Comment by marcospb19** [comment]
> If you could offer some guidance, I can try and put together a PR if you'd like?

Sorry for the late response. Unfortunately, I don't think I have the time to give guidance on how to approach this, but we could iterate on the solution in the PR review process.

You're right, `Ouch` doesn't support `--xattrs` atm.

---

I like your idea of having a `-P/--preserve-all-permissions` flag that covers it all.

In the `--help` message we also have to explain that these permissions apply for Tar but not for Zip archives :thinking:, and state that CLS is not supported yet.

In conclusion, I approve this feature request and a PR (from you or someone else) would be welcomed for reviewing :+1:.
